Source/DafnyLanguageServer/Workspace/DocumentDatabase.cs

```diff
@@ -134,36 +145,39 @@ IRelocator relocator
       var cancellationSource = new CancellationTokenSource();
-      var previousDocumentTask = databaseEntry.LatestDocument;
-      var resolvedDocumentTask = ApplyChangesAsync(previousDocumentTask, documentChange, cancellationSource.Token);
+      var updatedText = textChangeProcessor.ApplyChange(databaseEntry.TextBuffer, documentChange, CancellationToken.None);
+      var resolvedDocumentTask = ApplyChangesAsync(updatedText, databaseEntry, documentChange, cancellationSource.Token);
       var translatedDocument = LoadVerificationTasksAsync(resolvedDocumentTask, cancellationSource.Token);
-      var verifiedDocuments = Verify(translatedDocument, VerifyOnChange, cancellationSource.Token);
-      documents[documentUri] = new DocumentEntry(
+      var entry = new DocumentEntry(
         documentChange.TextDocument.Version,
+        updatedText,
         resolvedDocumentTask,
         translatedDocument,
-        verifiedDocuments,
-        cancellationSource
+        cancellationSource,
+        databaseEntry.Observer
       );
-      return resolvedDocumentTask.ToObservableSkipCancelled().
-        Concat(translatedDocument.ToObservableSkipCancelled()).
-        Concat(verifiedDocuments);
+      documents[documentUri] = entry;
+      entry.Observe(resolvedDocumentTask.ToObservableSkipCancelled()
+        .Concat(translatedDocument.ToObservableSkipCancelled()));
+
+      if (VerifyOnChange) {
+        Verify(entry, cancellationSource.Token);
+      }
     }
 
-    private async Task<DafnyDocument> ApplyChangesAsync(Task<DafnyDocument> oldDocumentTask, DidChangeTextDocumentParams documentChange, CancellationToken cancellationToken) {
-#pragma warning disable VSTHRD003
-      var oldDocument = await oldDocumentTask;
-#pragma warning restore VSTHRD003
+    private async Task<DafnyDocument> ApplyChangesAsync(DocumentTextBuffer updatedText, DocumentEntry documentEntry, DidChangeTextDocumentParams documentChange, CancellationToken cancellationToken) {
+
+      var oldDocument = documentEntry.LastPublishedDocument;
 
       // We do not pass the cancellation token to the text change processor because the text has to be kept in sync with the LSP client.
-      var updatedText = textChangeProcessor.ApplyChange(oldDocument.TextDocumentItem, documentChange, CancellationToken.None);
-      var oldVerificationDiagnostics = oldDocument.ImplementationViews ?? new Dictionary<ImplementationId, ImplementationView>();
+      var oldVerificationDiagnostics = oldDocument.ImplementationViewsView ?? new Dictionary<ImplementationId, ImplementationView>();
       var migratedImplementationViews = oldVerificationDiagnostics.ToDictionary(
         kv => kv.Key with {
           NamedVerificationTask =
           relocator.RelocatePosition(kv.Key.NamedVerificationTask, documentChange, CancellationToken.None)
         },
         kv => kv.Value with {
+          Range = relocator.RelocateRange(kv.Value.Range, documentChange, CancellationToken.None),
```

Comment:
I think here you should remove an implementation view if the ranged was deleted.